Create 142. Linked List Cycle II

mura0086 opened the pull request:
[142. Linked List Cycle II](https://leetcode.com/problems/linked-list-cycle-ii/)

問題文
Given the head of a linked list, return the node where the cycle begins. If there is no cycle, return null.

There is a cycle in a linked list if there is some node in the list that can be reached again by continuously following the next pointer. Internally, pos is used to denote the index of the node that tail's next pointer is connected to (0-indexed). It is -1 if there is no cycle. Note that pos is not passed as a parameter.

Do not modify the linked list.

mura0086 committed:
Create 142. Linked List Cycle II

oda commented on 142. Linked List Cycle II:
````diff
@@ -0,0 +1,75 @@
+ここでfastをスタート地点に移動させ１つずつポインタを進めるとslowとfastがサイクルの開始地点で重なることになる。
+（L-c+c=L=0）
+
+```cpp
````
「コードの整え方」を読んでおいてください。
https://docs.google.com/document/d/11HV35ADPo9QxJOpJQ24FcZvtvioli770WWdZZDaLOfg/edit?tab=t.0#heading=h.9kpbwslvv3yv

あとはこのへんかしら。
https://docs.google.com/document/d/11HV35ADPo9QxJOpJQ24FcZvtvioli770WWdZZDaLOfg/edit?tab=t.0#heading=h.jfs03xpyyrfl

mura0086 replied:
コメントありがとうございます。
それぞれ確認させていただきました。今後都度参照し選択肢を増やしていけるようにしたいと思います。
Tortoise and Hareについてはそういうものがあったな程度に留めておきます。

Fuminiton commented on 142. Linked List Cycle II:
```diff
@@ -0,0 +1,75 @@
+	    ListNode *detectCycle(ListNode *head) 
+	    {   
+	        ListNode *current_node = head;
+	        set<ListNode *> visited;
```
ご存知かもしれませんが、c++にはunordered_setなるものがあるそうです。
unordered_setで解いている方もいたので、参考になるかもしれません。

- [解答例](https://github.com/fhiyo/leetcode/pull/1/files)
- [公式ドキュメント](https://en.cppreference.com/w/cpp/container/unordered_set)
- [setとunorderd_setに関する議論](https://stackoverflow.com/questions/1349734/why-would-anyone-use-set-instead-of-unordered-set)

mura0086 replied:
コメント、参考リンクまでありがとうございます。計算量に余裕がある際はsetを使うようにしてますが、他にも理由を都度つけれるようにリンク内容ほか確認します。

quinn-sasha commented on 142. Linked List Cycle II:
```diff
@@ -0,0 +1,75 @@
+	public:
+	    ListNode *detectCycle(ListNode *head) 
+	    {   
+	        ListNode *current_node = head;
```
変数名についてですが、`current_node`は`prev`など対比するものがある場合に使われていると思います。
この場合、単に`node`としても良さそうですね。

mura0086 replied:
コメントありがとうございます。確かに、この問題では現在のノードからの参照しかないのでnodeでも良かったです。選択肢が増えました、ありがとうございます。

mura0086 committed:
Update 142. Linked List Cycle II

レビュー受けて、以下を変更
・STEP1のnode変数名を変更
current_node->node
時制を相対的に表現する必然性がなかったので単にnodeとするほうが感覚的にしっくりきた。

・フロイドの循環検出コードのスタート地点開始のロジックを関数化。（コードの整え方）
